src/importer.cpp

```diff
@@ -0,0 +1,484 @@
+        clearComponentImports(model->component(c));
+    }
+}
+
+void flattenComponent(const ComponentEntityPtr &parent, const ComponentPtr &component, size_t index)
+{
+    if (component->isImport()) {
+        auto model = owningModel(component);
+        auto importSource = component->importSource();
+        auto importModel = importSource->model();
+        auto importedComponent = importModel->component(component->importReference());
+
+        // Determine names of components already in use.
+        NameList compNames = componentNames(model);
+
+        // Determine the stack for the destination component.
+        IndexStack destinationComponentBaseIndexStack = reverseEngineerIndexStack(component);
+
+        // Determine the stack for the source component.
+        IndexStack importedComponentBaseIndexStack = reverseEngineerIndexStack(importedComponent);
+
+        // Generate equivalence map for the source component.
+        EquivalenceMap map;
+        recordVariableEquivalences(importedComponent, map, importedComponentBaseIndexStack);
+        generateEquivalenceMap(importedComponent, map, importedComponentBaseIndexStack);
+
+        // Rebase the generated equivalence map from the source component to the destination component.
+        auto rebasedMap = rebaseEquivalenceMap(map, importedComponentBaseIndexStack, destinationComponentBaseIndexStack);
+
+        // Take a copy of the imported component which will be used to replace the import defined in this model.
+        auto importedComponentCopy = importedComponent->clone();
+        importedComponentCopy->setName(component->name());
+        for (size_t i = 0; i < component->componentCount(); ++i) {
+            importedComponentCopy->addComponent(component->component(i));
+        }
+
+        // Get list of required units from component's variables.
+        std::vector<UnitsPtr> requiredUnits = unitsUsed(importModel, importedComponentCopy);
+
+        // Add all required units to a model so referenced units can be resolved.
+        auto requiredUnitsModel = Model::create();
+        for (const auto &units : requiredUnits) {
+            // Cloning units present elsewhere so that they don't get moved by the addUnits function.
+            if (units->parent() == nullptr) {
+                requiredUnitsModel->addUnits(units);
+            } else {
+                auto cloned = units->clone();
+                requiredUnitsModel->addUnits(cloned);
+            }
+        }
+
+        // Make a map of component name to component pointer.
+        ComponentNameMap newComponentNames = createComponentNamesMap(importedComponentCopy);
+        for (const auto &entry : newComponentNames) {
+            std::string newName = entry.first;
+            size_t count = 1;
+            while (std::find(compNames.begin(), compNames.end(), newName) != compNames.end()) {
+                newName += "_" + convertToString(count++);
+            }
```

Comment:
I would replace with:
```
size_t count = 0;
while (std::find(compNames.begin(), compNames.end(), newName) != compNames.end()) {
    newName += "_" + convertToString(++count);
}
```

Reply:
Not really a fan of changing the value of count before using it ... but ... meh. 98f73054

Reply:
FWIW, this is what we do a few lines below... :)